executor/joiner.go

```diff
@@ -435,3 +435,313 @@ func (j *innerJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *ch
 
 func (j *innerJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
 }
+
+func newJoinerEx(ctx sessionctx.Context, joinType plannercore.JoinType,
+	outerIsRight bool, defaultInner []types.Datum, filter []expression.Expression,
+	lhsColTypes, rhsColTypes []*types.FieldType) joiner {
+	base := baseJoiner{
+		ctx:          ctx,
+		conditions:   filter,
+		outerIsRight: outerIsRight,
+		maxChunkSize: ctx.GetSessionVars().MaxChunkSize,
+	}
+	colTypes := make([]*types.FieldType, 0, len(lhsColTypes)+len(rhsColTypes))
+	colTypes = append(colTypes, lhsColTypes...)
+	colTypes = append(colTypes, rhsColTypes...)
+	base.selected = make([]bool, 0, chunk.InitialCapacity)
+	if joinType == plannercore.LeftOuterJoin || joinType == plannercore.RightOuterJoin {
+		innerColTypes := lhsColTypes
+		if !outerIsRight {
+			innerColTypes = rhsColTypes
+		}
+		base.initDefaultInner(innerColTypes, defaultInner)
+	}
+	switch joinType {
+	case plannercore.SemiJoin:
+		base.shallowRow = chunk.MutRowFromTypes(colTypes)
+		return &semiJoinerEx{base}
+	case plannercore.AntiSemiJoin:
+		base.shallowRow = chunk.MutRowFromTypes(colTypes)
+		return &antiSemiJoinerEx{base}
+	case plannercore.LeftOuterSemiJoin:
+		base.shallowRow = chunk.MutRowFromTypes(colTypes)
+		return &leftOuterSemiJoinerEx{base}
+	case plannercore.AntiLeftOuterSemiJoin:
+		base.shallowRow = chunk.MutRowFromTypes(colTypes)
+		return &antiLeftOuterSemiJoinerEx{base}
+	case plannercore.LeftOuterJoin:
+		base.chk = chunk.NewChunkWithCapacity(colTypes, ctx.GetSessionVars().MaxChunkSize)
+		return &leftOuterJoinerEx{base}
+	case plannercore.RightOuterJoin:
+		base.chk = chunk.NewChunkWithCapacity(colTypes, ctx.GetSessionVars().MaxChunkSize)
+		return &rightOuterJoinerEx{base}
+	case plannercore.InnerJoin:
+		base.chk = chunk.NewChunkWithCapacity(colTypes, ctx.GetSessionVars().MaxChunkSize)
+		return &innerJoinerEx{base}
+	}
+	panic("unsupported join type in func newJoiner()")
+}
+
+type semiJoinerEx struct {
+	baseJoiner
+}
+
+func (j *semiJoinerEx) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
```

Comment:
It's weird.. Although `xxJoinerEx`s also implements the `joiner` interface **literally**, but actually they are different things compared with `joiner`. In joinner, the function signature of `tryToMatch()` is:

```go
tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (bool, error)
```

But the `xxJoinerEx`'s `tryToMatch()` actually is:

```go
tryToMatch(inner chunk.Row, outers chunk.Iterator, chk *chunk.Chunk) (matched bool, err error)
```

The evilest thing is, you named the first parameter to `outer` and the second parameter to `inners` to make them looks like the same...

This makes the code hard to read and maintain. Please do not to it this way...

Reply:
I think we can modify the `joinner` interface this way:

```go
type joiner interface {
	// the following two functions are the old functions, just renamed.
	tryToMatchInners(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (bool, error)
	onMissMatchInners(outer chunk.Row, chk *chunk.Chunk)

	// the following two functions are what you implemented in this PR.
	tryToMatchOuters(inner chunk.Row, outers chunk.Iterator, chk *chunk.Chunk) (bool, error)
	onMissMatchOuters(inner chunk.Row, chk *chunk.Chunk)
}
```